Convert cloud key pair form to angular

dgutride opened the pull request:
Purpose or Intent
-----------------
Convert the Compute > Clouds > Key Pairs form to Angular.

<img width="892" alt="screen shot 2016-06-27 at 3 48 38 pm" src="https://cloud.githubusercontent.com/assets/7306953/16393442/d66cd88c-3c7e-11e6-8f3a-e60573f3c1d5.png">

This included modifying the controller to remove any @edit variables and retrieve cloud providers for the dropdown.  The create method was also changed to allow for a POST from the new Angular controller.  All other functionality should be identical to the way it currently works.

Steps for Testing/QA
--------------------
> No additional testing is necessary other than verifying the key pair add form works the same way it used to (with the exception of having the required validation on the client side.

chessbyte commented:
/cc @martinpovolny

AparnaKarve commented:
@dgutride Thanks for the PR and the quick turnaround!
Overall, looks great and as discussed, I'll be sure to do a detailed review soon.

martinpovolny commented on app/controllers/auth_key_pair_cloud_controller.rb:
```diff
@@ -79,16 +68,16 @@ def set_form_vars
+  # REST call for provider choices
+  def ems_form_choices
+    assert_privileges("auth_key_pair_cloud_new")
+    ems_choices = []
```
no need for this line

martinpovolny commented on app/controllers/auth_key_pair_cloud_controller.rb:
```diff
@@ -79,16 +68,16 @@ def set_form_vars
+  def ems_form_choices
+    assert_privileges("auth_key_pair_cloud_new")
+    ems_choices = []
+    ems_choices = ManageIQ::Providers::CloudManager.select { |ems| ems.class::AuthKeyPair.is_available?(:create_key_pair, ems) }.map do |ems|
```
probably use `do`.. `end` here for both blocks
on shorter terminla this looks strange and unclear

martinpovolny commented on app/controllers/auth_key_pair_cloud_controller.rb:
```diff
@@ -79,16 +68,16 @@ def set_form_vars
     end
+    render :json => {
+      :ems_choices => ems_choices
+    }
```
end here you can use just one line

martinpovolny commented on app/controllers/auth_key_pair_cloud_controller.rb:
```diff
@@ -105,22 +94,24 @@ def new
+      options = {}
+      options[:name] = params[:name]
+      options[:public_key] = params[:public_key] if params[:public_key]
+      options[:ems_id] = params[:ems_id] if params[:ems_id]
```
```
options = {
  :name => params[:name]
  :public_key => !!params[:public_key]
....
}
```
if you like it?

martinpovolny commented on app/controllers/auth_key_pair_cloud_controller.rb:
```diff
@@ -154,10 +145,13 @@ def create
+      options = {}
+      options[:name] = params[:name]
+      options[:public_key] = params[:public_key] if params[:public_key]
+      options[:ems_id] = params[:ems_id] if params[:ems_id]
```
here too

martinpovolny replied:
same 4 lines, extract method?

martinpovolny commented on app/views/auth_key_pair_cloud/_form.html.haml:
```diff
@@ -1,44 +1,53 @@
-                          "data-miq_observe" => {:interval => '.5', :url => url}.to_json)
+        %input.form-control{:type            => "text",
+                            :name            => "name",
+                            "id"             => "name",
```
do we need to mix string and symbol keys?

maybe for ng-mode, althoug I thing there's `_` to `-` conversion

surely not for `id`

martinpovolny commented on app/views/auth_key_pair_cloud/_form.html.haml:
```diff
@@ -1,44 +1,53 @@
+                  "",
+                  "ng-model"                    => "keyPairModel.ems",
+                  "ng-options"                  => "ems.name for ems in ems_choices",
+                  "id"                          => "ems_id",
```
cool that you have `id` everywhere, QE will hate us less

AparnaKarve replied:
@martinpovolny You're welcome!  My branch <a href=https://github.com/ManageIQ/manageiq/compare/master...AparnaKarve:angular_step_by_step_demo?expand=1>here</a> was used as a reference which was also used in our demo during our talk in Mahwah. 😉

dgutride commented:
@martinpovolny - I think I've addressed all of your comments (including using rubymine to clean up the double-quotes/single-quote/colon property inconsistencies in the haml).

AparnaKarve commented:
@dgutride Can you squash the last 2 Rubocop related commits?

AparnaKarve commented on app/views/auth_key_pair_cloud/_form.html.haml:
```diff
@@ -37,11 +37,15 @@
+                  "required"                    => "",
+                  "pf-select"                   => true)
+          :javascript
+            miqInitSelectPicker();
```
@dgutride You can remove this JS method and instead use the directive `selectpicker-for-select-tag` for the `select_tag`

AparnaKarve commented:
If possible (without rocking the boat too much) ;), can you squash these 2 commits 760a257b18705c17d6194d4a8cd51b32b8ad699f and 4630ba37d3e27769d8caccff5ff18b1fc338f559 (since the second one is more of an 'undo' action)

AparnaKarve commented on app/controllers/auth_key_pair_cloud_controller.rb:
```diff
@@ -200,6 +105,7 @@ def key_pair_save
                            :model => ui_lookup(:table => 'auth_key_pair_cloud')
                          }
       end
+
```
blank line? Can you remove it?

AparnaKarve commented on app/controllers/auth_key_pair_cloud_controller.rb:
```diff
@@ -239,6 +145,7 @@ def key_pair_save
           page.replace("flash_msg_div", :partial => "layouts/flash_msg")
         end
       end
+        
```
same here

AparnaKarve commented:
@dgutride This looks great! I also verified that it works well in the UI 👍 
Once you address the comments above, this should be good for a merge!

dgutride commented:
@AparnaKarve - I've tried to squash 760a257 and 4630ba3 (since the second one is more of an 'undo' action) but the squash command will only allow me to compress onto the previous commit.  Would you like me to flatten that entire set?

AparnaKarve commented:
@dgutride Let's discuss this on gitter on how to squash those 2 non-consecutive commits.

AparnaKarve commented on app/views/auth_key_pair_cloud/_form.html.haml:
```diff
@@ -1,44 +1,53 @@
+                  :required                    => '',
+                  'pf-select'                  => true)
+          :javascript
+            miqInitSelectPicker();
```
@dgutride Thought I made a comment here yesterday...but apparently not.
You can remove this `:javascript` block completely and instead use the directive `selectpicker-for-select-tag` for the `select_tag`. Also, you would not need the `pf-select` directive.

AparnaKarve commented:
@dclarizio Verified that the form works well in the UI.
So this is good for a merge.
